frontend/src/router/routeHandlers.ts

```diff
@@ -0,0 +1,105 @@
+import { FamRouteError, RouteErrorName } from '@/errors/FamCustomError';
+import { routeItems } from '@/router/routeItem';
+import AuthService from '@/services/AuthService';
+import {
+    fetchApplicationRoles,
+    fetchApplications,
+    fetchUserRoleAssignments,
+} from '@/services/fetchData';
+import { asyncWrap } from '@/services/utils';
+import {
+    isApplicationSelected,
+    selectedApplication,
+} from '@/store/ApplicationState';
+import { populateBreadcrumb } from '@/store/BreadcrumbState';
+import { setRouteToastError as emitRouteToastError } from '@/store/ToastState';
+import type { RouteLocationNormalized } from 'vue-router';
+
+/**
+ * This file should contain only the Vue router handler and necessary
+ * helpers for router's life-cycle methods (beforeEach, beforeEnter etc...)
+ *
+ * Vue router does not wrap javascript error nicely.
+ * PimeVue Toast (current version) also has limitation which it cannot be used
+ * outside of Vue <setup>. This is the case in router.
+ * So, here in case of custom error (javascript error), emit the error to a
+ * state for it to be handled elsewhere.
+ */
+
+// --- beforeEnter Route Handler
+
+const beforeEnterDashboardRoute = async (to: RouteLocationNormalized) => {
+    // Requires fetching applications the user administers.
+    await asyncWrap(fetchApplications());
+    const userRolesFetchResult = await asyncWrap(
+        fetchUserRoleAssignments(selectedApplication.value?.application_id)
+    );
+    Object.assign(to.meta, { userRoleAssignments: userRolesFetchResult.data });
+    return true;
+};
+
+const beforeEnterGrantUserPermissionRoute = async (
+    to: RouteLocationNormalized
+) => {
+    populateBreadcrumb([routeItems.dashboard, routeItems.grantUserPermission]);
+
+    const appRolesFetchResult = await asyncWrap(
+        fetchApplicationRoles(selectedApplication.value!.application_id)
+    );
+    if (appRolesFetchResult.error) {
+        emitRouteToastError(appRolesFetchResult.error);
+        return { path: routeItems.dashboard.path };
```

Comment:
Hi @ianliuwk1019 , I only changed here, let it return to dashboard when there is an error before enter grant access page, thanks!!

Reply:
Thanks for the change Catherine! This route (in theory) should already come "from" dashboard.
However, I was thinking the same thing, that this can (for now) solve the case that user change browser address.